mount/titus-mount-ceph.c

```diff
@@ -0,0 +1,284 @@
+	exit(1);
+}
+
+static inline int pidfd_open(pid_t pid, unsigned int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static inline int fsopen(const char *fs_name, unsigned int flags)
+{
+	return syscall(__NR_fsopen, fs_name, flags);
+}
+
+static inline int fsmount(int fsfd, unsigned int flags, unsigned int ms_flags)
+{
+	return syscall(__NR_fsmount, fsfd, flags, ms_flags);
+}
+
+static inline int fsconfig(int fsfd, unsigned int cmd, const char *key,
+			   const void *val, int aux)
+{
+	return syscall(__NR_fsconfig, fsfd, cmd, key, val, aux);
+}
+
+static inline int move_mount(int from_dfd, const char *from_pathname,
+			     int to_dfd, const char *to_pathname,
+			     unsigned int flags)
+{
+	return syscall(__NR_move_mount, from_dfd, from_pathname, to_dfd,
+		       to_pathname, flags);
+}
+
+#define E_fsconfig(fd, cmd, key, val, aux)                                     \
+	do {                                                                   \
+		if (fsconfig(fd, cmd, key, val, aux) == -1)                    \
+			mount_error(fd, key ?: "create");                      \
+	} while (0)
+
+static void configure_source(int fsfd, const char *source) {
+	E_fsconfig(fsfd, FSCONFIG_SET_STRING, "source", source, 0);
+}
+
+static void process_option(char *option, int fsfd)
+{
+	/* Splits up a k=v string and runs fsconfig on it.
+           We supply all the inputs here, so it is safe, but parsing things
+           like this is a little dangerous */
+	char *key, *value, *saveptr;
+	key = strtok_r(option, "=", &saveptr);
+	option = NULL;
+	value = strtok_r(option, "=", &saveptr);
+
+	char value2[256] = "";
+	strcpy(value2, value);
+	/* We have to tack on an extra '==' because the secret is base64 encoded,
+	 * and this very basic string parsing will treat those == as tokens and skip
+	 * past them */
+	if (strcmp(key, "secret") == 0) {
+		char base64suffix[] = "==";
+		strcat(value2, base64suffix);
+	}
+	fprintf(stderr,
+		"titus-mount-ceph: Setting filesystem mount option %s=%s\n",
+		key, value2);
+	E_fsconfig(fsfd, FSCONFIG_SET_STRING, key, value2, 0);
+}
+
+static void do_fsconfigs(int fsfd, char *options)
+{
+	char *str1, *token, *saveptr;
+	/* Mount options come in in the classic comma-separated key=value pairs
+	   we need to split them up and pass them in for fsconfig to handle one at a time */
+	for (str1 = options;; str1 = NULL) {
+		token = strtok_r(str1, ",", &saveptr);
+		if (token == NULL)
+			break;
+		process_option(token, fsfd);
+	}
+	/* This last FSCONFIG_CMD_CREATE fsconfig call actually creates the superblock */
+	E_fsconfig(fsfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
+}
+
+static void switch_namespaces(int pidfd)
+{
+	int ret;
+	ret = setns(pidfd, CLONE_NEWNET | CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns net / mount");
+		exit(1);
+	}
+}
+
+static int setup_fsfd_in_namespaces(int sk, int pidfd)
+{
+	int fsfd, ret;
+	ret = setns(pidfd, CLONE_NEWUSER);
```

Comment:
Dumb question, can you do setns(pidfd, CLONE_NEWNET | CLONE_NEWNS | CLONE_NEWUSER)?

Reply:
I'm not 100% sure, but I thought, no, that doesn't work, because NEWUSER requires forking, and doesn't take effect in the current process?

Reply:
Or are you asking, can I just set them all at once?

Yes, I can. Previously though I split it out, one for each namespace to make it easier to understand *why* it couldn't setns.